src/fault_log.rs

```diff
@@ -1,135 +1,78 @@
-    BroadcastDecoding,
-    /// `HoneyBadger` could not deserialize bytes (i.e. a serialized Batch)
-    /// from a given proposer into a vector of transactions.
-    BatchDeserializationFailed,
-    /// `DynamicHoneyBadger` received a key generation message with an invalid signature.
-    InvalidKeyGenMessageSignature,
-    /// `DynamicHoneyBadger` received a key generation message with an invalid era.
-    InvalidKeyGenMessageEra,
-    /// `DynamicHoneyBadger` received a key generation message when there was no key generation in
-    /// progress.
-    UnexpectedKeyGenMessage,
-    /// `DynamicHoneyBadger` received a signed `Ack` when no key generation in progress.
-    UnexpectedKeyGenAck,
-    /// `DynamicHoneyBadger` received a signed `Part` when no key generation in progress.
-    UnexpectedKeyGenPart,
-    /// `DynamicHoneyBadger` received more key generation messages from the peer than expected.
-    TooManyKeyGenMessages,
-    /// `DynamicHoneyBadger` received a message (Accept, Propose, or Change)
-    /// with an invalid signature.
-    IncorrectPayloadSignature,
-    /// `DynamicHoneyBadger`/`SyncKeyGen` received an invalid `Ack` message.
-    SyncKeyGenAck(AckFault),
-    /// `DynamicHoneyBadger`/`SyncKeyGen` received an invalid `Part` message.
-    SyncKeyGenPart(PartFault),
-    /// `DynamicHoneyBadger` received a change vote with an invalid signature.
-    InvalidVoteSignature,
-    /// A validator committed an invalid vote in `DynamicHoneyBadger`.
-    InvalidCommittedVote,
-    /// `DynamicHoneyBadger` received a message with an invalid era.
-    UnexpectedDhbMessageEra,
-    /// `BinaryAgreement` received a duplicate `BVal` message.
-    DuplicateBVal,
-    /// `BinaryAgreement` received a duplicate `Aux` message.
-    DuplicateAux,
-    /// `BinaryAgreement` received multiple `Conf` messages.
-    MultipleConf,
-    /// `BinaryAgreement` received multiple `Term` messages.
-    MultipleTerm,
-    /// `BinaryAgreement` received a message with an epoch too far ahead.
-    AgreementEpoch,
-}
+
+//! Each algorithm can propogate their faulty node logs upwards to a calling algorithm via
+//! `DistAlgorihm`'s `.handle_input()` and `.handle_message()` trait methods.
+
+pub use failure::Fail;
 
 /// A structure representing the context of a faulty node. This structure
 /// describes which node is faulty (`node_id`) and which faulty behavior
 /// that the node exhibited ('kind').
 #[derive(Clone, Debug, PartialEq)]
-pub struct Fault<N> {
+pub struct Fault<N, F: Fail> {
     /// The faulty node's ID.
     pub node_id: N,
     /// The kind of fault the node is blamed for.
-    pub kind: FaultKind,
+    pub kind: F,
 }
 
-impl<N> Fault<N> {
-    /// Creates a new fault, blaming `node_id` for the `kind`.
-    pub fn new(node_id: N, kind: FaultKind) -> Self {
+impl<N, F> Fault<N, F>
+where
+    F: Fail,
+{
+    /// Creates a new fault given a node ID and a fault description.
```

Comment:
Minor, but I really liked the old docs because it made it abundantly clear what is going on (assigning blame for `kind` to `node_id`). Is there a reason to change it?